### PR TITLE
chore(compute): update_discovery_doc.sh now edits generator_config

### PR DESCRIPTION
**Pull request by scotthart**
Improve automation by having script edit generator_config.textproto when new compute services are detected and re-running the generator to emit code for the new services.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14623)

### CODE REVIEW

File: generator/discovery/update_discovery_doc.sh [hunk 1]
```diff
@@ -79,14 +89,27 @@ if [[ -n "${NEW_FILES}" ]]; then
...
   mapfile -t proto_array < <(echo "${NEW_FILES}")
-  io::log_red "Add rest_services definitions to the generator_config.textproto and re-run this script."
+  io::log_red "Adding rest_services definitions to the generator_config.textproto."
   for i in "${proto_array[@]}"; do
     print_service_textproto "${i}"
   done
   git add protos
-  io::log_yellow "Adding new directories to google/cloud/compute/service_dirs.cmake"
+
+  io::log_h2 "Running generate-libraries again with updated generator_config.textproto"
```
Comment (dbolduc) [inline generator/discovery/update_discovery_doc.sh:98]: > again

If `UPDATED_DISCOVERY_DOCUMENT` is set, we could cut off `generate-libraries.sh` after it writes the compute protos.

Then we can run  `build.sh -t generate-libraries-pr || true` here without setting `UPDATED_DISCOVERY_DOCUMENT`?

It seems cleaner to do all the `git add`ing in this script.

File: generator/discovery/update_discovery_doc.sh [hunk 2]
```diff
@@ -23,7 +23,10 @@ function print_service_textproto() {
   service_proto_path="${1#protos/}"
   product_path="${service_proto_path%/*.proto}"
   initial_copyright_year=$(date +"%Y")
-  cat <<_EOF_
+
+  (
```
Comment (dbolduc) [inline generator/discovery/update_discovery_doc.sh:27]: I am too lazy to figure out what these commands do. My guess is that it fills in the text between "Begin" and "End" with the new additions. But doesn't that get overwritten the next time we update the discovery doc?

Also do we care about sorting?

---

A solution where we have all the `rest_services {}` between "Begin" and "End" is close. The issue is with the copyright year which is variable. Ugh.

If we don't care about sorting, we can have one marker, and just insert the new `rest_services {}` above it in the config.

**Review by dbolduc — CHANGES_REQUESTED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14623)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @dbolduc)

___
*[`generator/discovery/update_discovery_doc.sh` line 27 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14623#gh-1713112688:-O46J_U_2uTIWEE4lzFM:b-1ii3da) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/76fcfd6058fb2fd26e4f90fb82770caf4e964cdd/generator/discovery/update_discovery_doc.sh#L27)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

I am too lazy to figure out what these commands do. My guess is that it fills in the text between "Begin" and "End" with the new additions. But doesn't that get overwritten the next time we update the discovery doc?

Also do we care about sorting?

---

A solution where we have all the `rest_services {}` between "Begin" and "End" is close. The issue is with the copyright year which is variable. Ugh.

If we don't care about sorting, we can have one marker, and just insert the new `rest_services {}` above it in the config.
</blockquote></details>

Fixed the `sed` stuff to append new service definitions instead of overwriting. The sorting was mostly for benefit of the human editing it, but this should remove that interaction. One marker mechanism adopted.
___
*[`generator/discovery/update_discovery_doc.sh` line 98 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14623#gh-1713112438:-O47KsY22UYopx21EKq_:b-gijd9x) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/76fcfd6058fb2fd26e4f90fb82770caf4e964cdd/generator/discovery/update_discovery_doc.sh#L98)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

> again

If `UPDATED_DISCOVERY_DOCUMENT` is set, we could cut off `generate-libraries.sh` after it writes the compute protos.

Then we can run  `build.sh -t generate-libraries-pr || true` here without setting `UPDATED_DISCOVERY_DOCUMENT`?

It seems cleaner to do all the `git add`ing in this script.
</blockquote></details>

Updated generate-libraries to exit early after emitting protos (when UPDATED_DISCOVERY_DOCUMENT is set). Moved all the `git add` from generate-libraries to this script.

File: generator/discovery/update_discovery_doc.sh [hunk 3]
```diff
@@ -69,24 +78,60 @@ readonly REVISION
...
+
+if [[ -n "${NEW_FILES}" ]]; then
+  io::log_h2 "Adding new compute generated service code"
+  git add google/cloud/compute
+
+  io::log_h2 "Formatting generated code"
+  git ls-files -z -- '*.h' '*.cc' |
+    xargs -P "$(nproc)" -n 50 -0 clang-format -i
+  git ls-files -z -- '*.h' '*.cc' |
+    xargs -r -P "$(nproc)" -n 50 -0 sed -i 's/[[:blank:]]\+$//'
```
Comment (dbolduc) [inline generator/discovery/update_discovery_doc.sh:121]: nit: we are not pressed for time or anything, but we could scope these to `google/cloud/compute`

**Review by dbolduc — APPROVED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14623)* status: 0 of 3 files reviewed, 3 unresolved discussions (waiting on @dbolduc)

___
*[`generator/discovery/update_discovery_doc.sh` line 121 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14623#gh-1715503511:-O4BPrDR9RBawntLA0S4:b-f6qlzn) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/021d22226753f20e043a0a920bfd4bfe17084023/generator/discovery/update_discovery_doc.sh#L121)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: we are not pressed for time or anything, but we could scope these to `google/cloud/compute`
</blockquote></details>

I'll look into that next time I'm modifying this script.

**Pull request merged**